components/script/dom/document.rs

```diff
@@ -1556,11 +1556,14 @@ impl Document {
                 return;
             }
 
+            // For a node within a text input UA shadow DOM, redirect the focus target into its shadow host.
+            let target_el = el.find_focusable_shadow_host_if_necessary();
```

Comment:
I think a better approach is to make the inner editor focusable, and set `delegatesFocus` on the shadow root

https://developer.mozilla.org/en-US/docs/Web/API/Element/attachShadow#delegatesfocus

Do we support that currently?

Reply:
no, we do not support `delegateFocus`. Our focus implementation as a whole deviates from the spec significantly, there's a todo in `Element::Focus` about that.